portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -1280,6 +1479,101 @@ else if (isBlankValue(newValues)) {
 	private boolean _windowStateRestoreCurrentView;
 	private String _windowStateString;
 
+	private class PortletURLParameterMap extends AbstractMap<String, String[]> {
```

Comment:
@shuyangzhou I recommend that you start your review of PortletURLImpl.java here with the `PortletURLParameterMap` inner class. Standby, I will write more information shortly.

Reply:
@shuyangzhou The reason why `PortletURLParameterMap` exists is because the `PortletURL.getParameter(String)` methods and `PortletURL.setParameter(...)` methods  have been deprecated in the Portlet 3.0 API. Those methods have been replaced by the composition objects returned by `PortletURL.getActionParameters()`, `PortletURL.getRenderParameters()`, and `PortletURL.getResourceParameters()`.

Reply:
@shuyangzhou During development I had two choices:

1. Have the deprecated stuff call-through to the new stuff.
2. Have the new stuff call-through to the deprecated stuff.

I chose #1. And that's why the `PortletURLParameterMap` inner class exists. As you can see, it extends `AbstractMap` and the `entrySet()` method is a generalized layer of abstraction over the clear distinction between action parameters, render parameters, and resource parameters.